Library/Homebrew/cmd/uninstall.rb

```diff
@@ -26,6 +53,36 @@ def uninstall
         name = Formulary.canonical_name(name)
         rack = HOMEBREW_CELLAR/name
 
+        # Cellar/oldname can be a symlink to Cellar/newname so in order not to
+        # get newname's keg as oldname's keg we check if it's a symlink.
+        if rack.symlink?
+          raise "#{rack.to_s} is a symlink"
+        end
```

Comment:
I would like to treat old name as an alias of the formula. And `canonical_name` should always return **canonical** name(i.e. not old name).

I don't want to change `Dependency#to_formula` method because there are much deeper problems should be addressing. By altering that method, you make existing problems more complicate.